replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/modern/AbstractPersistable.java

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ditto.replication.api.impl.modern;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import javax.annotation.Nullable;
+import org.codice.ditto.replication.api.modern.Persistable;
+
+/**
+ * This class defines some data that any object to be saved in persistence should have and methods
+ * to manipulate that data.
+ */
+public abstract class AbstractPersistable implements Persistable {
+
+  private static final String ID = "id";
+
+  private String type;
+
+  protected Map<String, Object> properties;
+
+  protected AbstractPersistable(@Nullable String id, String type) {
+    this.type = type;
+    properties = new HashMap<>();
+
+    if (id == null) {
+      setId(UUID.randomUUID().toString());
+    } else {
+      setId(id);
+    }
+  }
+
+  protected AbstractPersistable(String type, Map<String, Object> properties) {
+    this.type = type;
+    setProperties(properties);
+    if (properties.get(ID) == null) {
+      setId(UUID.randomUUID().toString());
+    }
+  }
+
+  public String getId() {
+    return (String) properties.get(ID);
+  }
+
+  public void setId(String id) {
+    safePut(ID, id);
+  }
+
+  public String getType() {
+    return type;
+  }
+
+  public void setType(String type) {
+    this.type = type;
+  }
+
+  public Map<String, Object> getProperties() {

```

Comment:
❗️ This should probably be package private since it is only required by the persistence manager and therefore could return the map as is as opposed to cloning it. Worst case scenario, could you not just return an unmodifiable version of it?